---
title: "CSAMA2019: Mass spectrometry and proteomics lab"
author:
- name: Laurent Gatto
  affiliation: Computational Biology group, de Duve Institute, UCLouvain, Belgium
- name: Lieven Clement
  affiliation: statOmics group, Gent University, Belgium
date: "`r date()`"
output:
  BiocStyle::html_document
---

# Introduction

This tutorial requires  the
`r BiocStyle::Biocpkg("MSnbase")` ([Gatto and Lilley
(2011)](https://academic.oup.com/bioinformatics/article/28/2/288/199094)),
`r BiocStyle::Githubpkg("statOmics/MSqRob")` ([Goeminne *et al.*
(2016)](https://www.ncbi.nlm.nih.gov/pubmed/26566788)),
`r BiocStyle::Biocpkg("limma")` ([Ritchie *et al.*
 (2015)](https://academic.oup.com/nar/article/43/7/e47/2414268)) and
`r BiocStyle::Biocexptpkg("msdata")` packages, as well as some of the
`r BiocStyle::CRANpkg("tidyverse")`
packages for data tidying and visualisation.

```{r env, message = FALSE}
library("MSnbase")
library("MSqRob")
library("msdata")
library("limma")
library("tidyverse")
```

Sections 2 to 4 deal with raw mass spectrometry and identification
data stored in one of the open community-developed data formats (see
below). Section 5 describes how to produce or import quantitative data
and the sections thereafter focus on the processing and analysis of
differentially expressed proteins.

### Data files in mass spectrometry  {.unnumbered}

Most community-driven formats described in the table are supported in
`R`. We will see how to read and access these data in the following
sections.

```{r datatab, results='asis', echo=FALSE}
datatab <-
    data.frame(Type = c("raw", "identification", "quantitation",
                   "peak lists", "quantitation (with id)", "protein db"),
               Format = c("mzML, mzXML, netCDF, mzData",
                   "mzIdentML", "mzQuantML", "mgf", "mzTab, spreadsheets", "fasta"),
               Package = c(
                   "*[MSnbase](http://bioconductor.org/packages/MSnbase)* (read and write in version >= 2.3.13) via *[mzR](http://bioconductor.org/packages/mzR)*",
                   paste("*[mzID](http://bioconductor.org/packages/mzID)* (read) and",
                         "*[MSnbase](http://bioconductor.org/packages/MSnbase)* (read, via *[mzR](http://bioconductor.org/packages/mzR)*)"),
                   "",
                   "*[MSnbase](http://bioconductor.org/packages/MSnbase)* (read)",
                   "*[MSnbase](http://bioconductor.org/packages/MSnbase)* (read, read/write)",
                   "*[Biostrings](http://bioconductor.org/packages/Biostrings)*"))

knitr::kable(datatab)
```

In the next sections, we will use raw data (in `mzML` and `mzXML`
format), identification data (in `mzIdentML` format) and quantitative
data (from a tab-separated spreadsheet).

# Raw MS data

Raw data files (in any of the above formats) is read into R using
`readMSData` function from the `MSnbase` package, that will return an
list-like object of class `MSnExp`. Below, we first extract the full
path to the `MS3TMT11.mzML` file from the `msdata` package^[The
`proteomics`, `ident` and `quant` `msdata` functions return example
files for raw, identification and quantitative data respectively.]
before reading it in.


```{r data1}
basename(fl3 <- msdata::proteomics(full.name = TRUE, pattern = "MS3TMT11"))
(rw3 <- readMSData(fl3, mode = "onDisk"))
```

Note: above, we specify to use the *on disk* mode (as opposed to *in
memory*) to avoid loading the whole raw data file(s) into memory. With
*on disk*, the raw data will be accessed on demand. Here, we only read
a single MS file into R, but with *on disk* mode, we could load 100s
thereof.

An `MSnExp` object contains 

- the raw data (spectra), accessible with `[` and `[[`. The former
  returns and subset of the data as a new `MSnExp` object, while the
  former extracts a single spectrum.
- features data, accessible as a dataframe with `fData`

**Exercise**: Extract the first, second and tenth spectrum. What are
their MS levels, precursor m/z and retention times?

<details>
```{r ex0}
rw3[[1]]
```
</details>

We can also access specific data for the whole experiment using
accessors:

- `msLevel(rw3)`
- `rtime(rw3)`
- `precursorMz(rw3)`
- `centroided(rw3)`
- ...

See `?MSnExp` for details.


**Exercise**: Using `msLevel`, extract the MS level of the
`r length(rw3)` spectra of the `rw3` file. What levels are available in these
data. Using `centroided`, check what spectra and centroided of in
profile mode (see figure below). Which are the levels that are
centroided or in profile mode.


```{r centroided, fig.cap = "Peak picking. Raw profile mode data (left) as measured by the mass spectrometre, and processed, centroided data (right). See also this [`MSnbase` vignette](http://lgatto.github.io/MSnbase/articles/v03-MSnbase-centroiding.html).", echo = FALSE}
knitr::include_graphics("./Figures/centroided.png")
```


<details>

```{r ex1}
table(msLevel(rw3), centroided(rw3))
```
</details>

# Identification results

Identification data in `mzIdentML` is parsed and loaded as a
`data.frame` using the `readMzIdData` function.


```{r id}
basename(idf <- msdata::ident(full.name = TRUE))
iddf <- readMzIdData(idf)
```

Among the identification variables available, you will find
`MS.GF.RawScore`, the identification score computated by MSGF+ (the
search engine used for the search), and `isDecoy`. 

```{idvars}
names(iddf)
```

**Exercise**: How many PSMs from the read and decoy database are
recoded in the identification data? Calculate summary statistics for
MSGF+ raw scores.

<details>
```{r idex0}
table(iddf$isDecoy)
summary(iddf$MS.GF.RawScore)
```
</details>

**Exercise**: Reproduce and interpret the plot showing the
identification score distribution for the decoy and real peptides in
this file.

<details>

```{r fdrplot}
ggplot(iddf, aes(x = MS.GF.RawScore, colour = isDecoy)) +
    geom_density()
```
</details>


# Combining raw and identification data

When working with raw and identification data, these two can be merged
by adding the identification results to the raw feature meta-data
slot. Below, we use small data that is shipped with the `MSnbase`
package. The raw data being very small, we can afford to read it into
memory without specifying `mode = "onDisk"`.


```{r addid}
basename(quantFile <- dir(system.file(package = "MSnbase", dir = "extdata"),
                 full.name = TRUE, pattern = "mzXML$"))
basename(identFile <- dir(system.file(package = "MSnbase", dir = "extdata"),
                          full.name = TRUE, pattern = "dummyiTRAQ.mzid"))
     
msexp <- readMSData(quantFile)
fvarLabels(msexp)
```

Add this stage, there's only one `spectrum` feature variable with the
spectra indices. Below we add the identification data with
`addIdentificationData` to gain 33 new feature variables.


```{r addid2}
msexp <- addIdentificationData(msexp, identFile)
fvarLabels(msexp)
```

Note that prior to addition, the identification data is filtered as
documented in the `filterIdentificationDataFrame` function:

1. only PSMs matching the regular (non-decoy) database are retained;
2. PSMs of rank greater than 1 are discarded; 
3. only proteotypic peptides are kept.


# Quantitative data

Quantitative data is stored a objects of class `MSnSet`, that are
composed of an expression matrix (accessed with `exprs()`), a feature
meta-data dataframe (accessed with `fData()`) and a meta-data
dataframe (accessed with `pData()`).

```{r msnset, fig.cap = "The MSnSet structure", echo = FALSE}
knitr::include_graphics("./Figures/msnset.png")
```

`MSnSet` objects can be:

1. Created from an `MSnExp` object with the `quantify` function.
2. Created from a spreadsheet produced by a third party software using
   the `readMSnSet2` function.

## Quantifying raw data

```{r quant}
data(itraqdata)
itraqdata
msnset <- quantify(itraqdata, method = "trap", reporters = iTRAQ4)
msnset
```

Below, we plot the first MS2 spectrum from the `itraqdata` test data,
highlighting the four iTRAQ reporter ions and extract the quantiation
values and feature metadata of that same first record.

```{r qvis, fig.cap = "An iTRAQ spectrum. The full spectrum is the composite of the peptide from all samples and is used for identification. The reporter ion peaks are specific to each sample and are used for quantitation. "}
plot(itraqdata[[1]], reporters = iTRAQ4, full = TRUE)
exprs(msnset)[1, ]
fData(msnset)[1, ]
```

## Importing third-party data

We are going to use the `cptac_a_b_peptides.txt` tab-separated file
from the `msdata` package. These data are the 6th study of the
Clinical Proteomic Technology Assessment for Cancer (CPTAC). In this
experiment, the authors spiked the Sigma Universal Protein Standard
mixture 1 (UPS1) containing 48 different human proteins in a protein
background of 60 ng/μL Saccharomyces cerevisiae strain BY4741 . Two
different spike-in concentrations were used: 6A (0.25 fmol UPS1
proteins/μL) and 6B (0.74 fmol UPS1 proteins/μL). We limited ourselves
to the data of LTQ-Orbitrap W at site 56. The data were searched with
MaxQuant version 1.5.2.8, and detailed search settings were described
in [Goeminne *et al.*
(2016)](https://www.ncbi.nlm.nih.gov/pubmed/26566788). Three
replicates are available for each concentration. The study is a
spike-in study for which we know the ground truth so we have the
ability to evaluate the quality of the fold change estimates and the
list of DE genes that we return with a method.


```{r cptacfile}
basename(f <- msdata::quant(full.names = TRUE))
```

Before reading the spreadsheet, we need to identify which columns
contain quantitation data (that will be used to populate the `exprs`
slot) and the feature data (that will be put into the `fData` slot).

The `getEcols` function lists the column names in the expression data
spreadsheet. The quantitative values we want to used are those in the
columns starting with`Intensity 6A_7`, `Intensity 6A_8`, ... 
`Intensity 6B_9`, that we refer to with `Intensity `. 


```{r read0}
getEcols(f, split = "\t")
```

Using a pattern, we can set the columns to be used to populate the
quantitation slot.


```{r read}
(e <- grepEcols(f, "Intensity ", split = "\t")) ## careful at the space!
(cptac <- readMSnSet2(f, ecol = e,
                      fnames = "Sequence",
                      sep = "\t"))
```


We can access the peptide-level expression data with `exprs` and the
feature meta-data with `fData`.

```{r exprs_and_fData}
head(exprs(cptac))
tail(exprs(cptac))
fvarLabels(cptac)
```

For the sake of simplicity, we can clean up the feature variables and
only keep those of interest. It is possible to do this interactively
with

```{r selectfeats0, eval = FALSE}
cptac <- selectFeatureData(cptac)
```

or by setting the feature variables of interest.

```{r selectfeats}
cptac <- selectFeatureData(cptac,
                           fcol = c("Proteins",
                                    "Potential.contaminant",
                                    "Reverse",
                                    "Sequence"))
fvarLabels(cptac)
```

Let's also add sample annotations:

```{r pdata}
cptac$condition <- factor(rep(c("A", "B"), each = 3))
cptac$sample <- rep(7:9, 2)
pData(cptac)
```

This could also be done by reading a spreadsheet into R as a
`data.frame`, making sure that the rownames match the sample names
exactly, and then adding it with `pData(cptac) <- myDf`.


The sample names are rather long and contain information on the
spike-in concentration and the repeat. We this remove `Intensity.6`
from the sample names:

```{r snms}
sampleNames(cptac) <- sub("Intensity\\.6", "", sampleNames(cptac))
sampleNames(cptac) 
```

Note that data in the `mzTab` format can be imported using the
`readMzTabData` function.

# Quantitative data processing

## Handling overlapping protein groups

In our approach a peptide can map to multiple proteins, as long as
there is none of these proteins present in a smaller subgroup.

```{r smugr0}
head(keep <- smallestUniqueGroups(fData(cptac)$Proteins))
length(keep)
length(unique(fData(cptac)$Proteins))
```

As shown above, we will keep `r length(keep)` protein groups
from the `r length(unique(fData(cptac)$Proteins))` 
protein groups recorded in the data. This vector of protein names is
used to filter the peptide-level data.

```{r smugr1}
(cptac <- cptac[fData(cptac)$Proteins %in% keep, ])
```

**Exercise**: How many peptides have we lost by removing the proteins
above?

## Filtering out contaminants and reverse hits

Below, we create vector of logicals (and count) recording peptides
that are assigned to contaminant (such as keratine, trypsine, ...) and
reverse proteins (from the decoy database). These are annotated with a
`"+"` in the respective `"Potential.contaminant"` and `"Reverse"`
feature variables.


```{r conts}
table(sel_conts <- fData(cptac)$Potential.contaminant != "+")
```

```{r rev}
table(sel_rev <- fData(cptac)$Reverse != "+")
```

```{r filtering}
(cptac <- cptac[sel_conts & sel_rev, ])
```

You can keep track of the data processing steps in the object's
processing log.

## Removing PTM-only proteins

Proteins for which all peptides are carrying modifications (PTMs) can
be considered as unreliable. We will filter out these proteins. This
information is included in the `Only.identified.by.site` column of the
`proteinGroups.txt` MaxQuant file. The code chunk below provides this
list of proteins.

```{r ptm_only}
ptm_only <- c("REV__CON__Q3T052", "REV__sp|P01120|RAS2_YEAST",
              "REV__sp|P32849|RAD5_YEAST",
              "REV__sp|Q03723|OST6_YEAST", "sp|P04051|RPC1_YEAST",
              "sp|P06367|RS14A_YEAST",
              "sp|P0CX73|YP11A_YEAST;sp|P0CX72|YL12A_YEAST;sp|P0CX71|YE11A_YEAST;sp|P0CX70|YD15A_YEAST;sp|Q6Q5H1|YP14A_YEAST;sp|P0C2I8|YL14A_YEAST",
              "sp|P19657|PMA2_YEAST", "sp|P32465|HXT1_YEAST",
              "sp|P39567|IMDH1_YEAST", "sp|P40527|ATC7_YEAST",
              "sp|P40530|PDK1_YEAST", "sp|P40989|FKS2_YEAST",
              "sp|P49955|SF3B1_YEAST", "sp|P51401|RL9B_YEAST",
              "sp|P53072|TAN1_YEAST", "sp|Q03964|YD17A_YEAST",
              "sp|Q04670|YM14B_YEAST;sp|Q12088|YL11B_YEAST;sp|Q03619|YE12B_YEAST",
              "sp|Q08649|ESA1_YEAST", "sp|Q12112|YN11B_YEAST",
              "sp|Q12479|IRC11_YEAST", "sp|Q3E7B7|YD85C_YEAST")
```

We now remove the peptides matched to these proteins:

```{r ptm_only_remove}
(cptac <- cptac[!fData(cptac)$Proteins %in% ptm_only, ])
```

## Notes on missing values

Unfortunately, some software use 0 irrespective whether the data has
intensity zero and when the data haven't been observer. Below we fix
this by setting all 0 values to `NA`.


```{r setna}
exprs(cptac)[exprs(cptac) == 0] <- NA
table(is.na(exprs(cptac)))
```

The following figure shows the distribution of missing values for
samples (columns) and rows (peptides). The cells have been reporder to
emphasis the presence of missing values in the proteins shown at the
top and the samples shown towards the right.

```{r napac, fig.cap = "Overview of missing data"}
napac <- cptac
exprs(napac)[!is.na(exprs(napac))] <- 1
naplot(napac)
```

The following figure the proportions of features (peptides in this
case) with respect to their completeness (blue) and the percentage of
missing data in the full dataset (red).

```{r plotNA, fig.cap = "Data completeness."}
plotNA(cptac)
```

Below, we count the number of missing values in each 

```{r nna}
fData(cptac)$nNA <- apply(exprs(cptac), 1, function(x) sum(is.na(x)))
table(fData(cptac)$nNA)
```

Note that some peptides aren't seen at all because these 6 samples are
a subset of a larger dataset, and these features are present in the
other acquisitions only.

From here on one could:

- **filter** data with missing values, which however sacrifices a lot
  of data.

```{r filterNA}
filterNA(cptac)
```

- perform **imputation**, considering the underlying nature of
  missingness, i.e missing not at random (left-censored) or at random.


```{r lazar, fig.cap = "Effect of the nature of missing values on their imputation. Root-mean-square error (RMSE) observations standard deviation ratio (RSR), KNN and MinDet imputation. Lower (blue) is better.", echo = FALSE}
knitr::include_graphics("./Figures/imp-sim.png")
```

See Lazar *et al.* [Accounting for the Multiple Natures of Missing
Values in Label-Free Quantitative Proteomics Data Sets to Compare
Imputation
Strategies](http://dx.doi.org/10.1021/acs.jproteome.5b00981).

- The best solution is arguably to handle missing values at **the
  statistical test level**, which is the approach we are going to use.

We are going to keep peptides that had a least two observations. This
can be done with the `nNA` variable that we compute above

```{r nNAfilt}
(cptac <- cptac[fData(cptac)$nNA <= 4, ])
```

## Log transformation

```{r log}
(cptac <- log(cptac, base = 2))
```

## Normalisation

Normalisation is handled by the `normalise` (or `normalize`) function.

```{r norm}
plotDensities(exprs(cptac))
(cptac <- normalise(cptac, method = "quantiles"))
plotDensities(exprs(cptac))
```

We can visualise our peptide-level data using a Multi Dimensional
Scaling (MDS) plot, using the `plotMDS` function from the `limma`
package. We use the condition sample variable to colour-code the
samples.


```{r mds0}
plotMDS(exprs(cptac), col = as.numeric(cptac$condition))
```

The first axis in the plot is showing the leading log fold changes
(differences on the log scale) between the samples.  We notice that
the leading differences in the peptide data seems to be driven by
technical variability. Indeed the samples do not seem to be clearly
separated according to the spike in condition.

## Summarisation

So far we have used quantitation values at the peptide level, while
the data of interest are proteins. We can take all peptides that are
associated with a protein group, as defined by the `Proteins` feature
variable, and aggregate them using an summary function of choice.


```{r featsfig2, fig.cap = "Multiple levels of summarisation. Examples of aggregation from PSMs, peptides to proteins using median aggregation (from the [`Features`](https://rformassspectrometry.github.io/Features/articles/Features.html) package).", echo = FALSE}
knitr::include_graphics("./Figures/features.png")
```

Below, we combine the peptides into proteins using the median, passing
`na.rm = TRUE` to account for the missing values in the data. 

```{r combine_med}
(cptac_prot <- combineFeatures(cptac, fcol = "Proteins",
                               method = "median", na.rm = TRUE))
```

We obtain `r nrow(cptac_prot)` proteins. Note how the processing steps are recorded. Below, we
visualise the protein-level data on an MDS plot.


```{r mds1}
plotMDS(exprs(cptac_prot), col = as.numeric(cptac_prot$condition))
```

**Exercise**: Repeat the summarisation using a the robust
normalisation by setting`method = "robust"`, as described in [Sticker *et
al*. (2019)](https://www.biorxiv.org/content/10.1101/668863v1) and
visualise the data on an MDS plot. Which one do you anticipate to
provide better results?

<details>
```{r combine_rob, warning = FALSE}
cptac_rob <- combineFeatures(cptac, fcol = "Proteins",
                             method = "robust", na.rm = TRUE)
plotMDS(exprs(cptac_rob), col = as.numeric(cptac_rob$condition))
```
</details>


# Differential expression analysis

**Exercise**: In this section, the median-summarised protein data will
be used. As an exercice, you are advised to use the best of the
median- or robust-summerised data and compare your results with those
shown below.

## Estimation

`MSqRob` is currently working with a format where we have one
dataframe for each protein.  This will be changed in the next release
to use the `MSnSet` directly. Therefore we first have to reorganise
the data.

```{r dfprot, cache = TRUE}
protMSqRob <- MSnSet2protdata(cptac_prot, "Proteins")
```

Next the models are fitted. This is done using the `fit.model`
function.  We only have to model the data using the factor condition
from the sample metadata (the `pData` slot) of the protein level
`MSnSet`.  The name of the factor variable is specified in the fixed
argument (if multiple predictors have to be incorporated in the model,
a vector of variable names has to be provided in this argument). The
argument shrinkage is used to specify if ridge regression has to be
adopted.  For the sake of speed we do not do this in the tutorial. The
shrinkage has to be specified for each variable in the fixed
effects. We also have to indicate this for the intercept (which we
never shrink).  So we specify it at `c(0, 0)` to indicate that the
intercept (first 0) and the parameters for the factor condition
(second 0) are not penalized.  We set the `robust_var` function equal
to `FALSE` - this functionality will be removed from the package in
the next release.


```{r models, cache = TRUE}
models <- fit.model(protdata = protMSqRob,
                    response = "quant_value",
                    fixed = "condition",
                    shrinkage.fixed = c(0, 0),
                    robust_var = FALSE)
```

## Inference

Often, biologists have problems with the reference coding. In MSqRob
we have opted to formulate contrasts using all levels of a factor.
Internally, the contrasts are than recasted according to the factor
level that is the reference class.

```{r inference}
L <- makeContrast("conditionB - conditionA",
                  levels = c("conditionA", "conditionB"))
res <- test.contrast_adjust(models, L)
head(res)
```

Below, we put the results of the statistical analysis back into the
`MSnSet` feature data, to keep the data and their analysis
together. We need however to rearrange the proteins in the result
dataframe (ordered by adjusted p-value) to match the order in the
`MSnSet`.


```{r inf_in_msnset}
fData(cptac_prot)$res <- res[featureNames(cptac_prot), ]

```

There are `r sum(fData(cptac_prot)$res$signif, na.rm = TRUE)` 
protein groups identified as differentially expressed at a significant
effect at the 5% FDR level.


# Visualising DE results

## Volcano plot

```{r volc}
volc <- ggplot(fData(cptac_prot)$res,
               aes(x = estimate,
                   y = -log10(pval),
                   color = signif)) +
    geom_point() +
    scale_color_manual(values = c("black", "red"))
volc
```

It is easy to generate an interactive graph to explore the results
using the `plotly` package and passing the ggplot-object to the
`ggplotly` function:

```{r volcly, eval = FALSE}
plotly::ggplotly(volc)
```

## Heatmap

```{r hmap}
sign <- which(fData(cptac_prot)$res$sig)
heatmap(exprs(cptac_prot)[sign,])
```

## Expression data for proteins of interest

Below, we extract the peptide data matching the differentially
expressed proteins.

```{r de_peps}
sign_prots <- featureNames(cptac_prot)[sign]
(cptac_sign <- cptac[fData(cptac)$Proteins %in% sign_prots, ])
```

In the following code chunk, we convert the peptide-level `MSnSet`
containing the significant peptides using the `ms2df` helper function,
convert that wide format dataframe to a long format and visualise the
expression distributions in each group.

```{r de_peps_ggplot, warning = FALSE}
ms2df(cptac_sign) %>%
    tidyr::gather(key = sample, value = expression, 1:6) %>%
    ggplot(aes(x = sample, y = expression)) +
    geom_boxplot() +
    geom_jitter(aes(colour = Sequence)) +
    facet_grid(Proteins ~ .)
```

## Global fold-changes

Because we know the ground truth for the cptac study, i.e. we know
that only the spike-in proteins (UPS) are differentially expressed, we
can evalute the fold changes. Yeast proteins should be not
differentially expressed and their log fold changes should be centered
around 0. These of UPS proteins are spiked at differt concentrations
and their log2 fold changes should be centered around $log2(concB/concA)$, i.e $log2(0.74/0.25) = 1.56$.

```{r lfc}
fData(cptac_prot)$res$spike <- grepl("UPS", fData(cptac_prot)$Proteins)

ggplot(fData(cptac_prot)$res,
       aes(x = spike, y = estimate)) +
    geom_boxplot() +
    ylab("log2 FC") + 
    geom_hline(yintercept = c(0, log(0.74/0.25, base = 2)),
               color = "red")
```

## Sensitivity FDP plot

Because we know the ground truth for the cptac study, i.e. we know
that only the spike-in proteins (UPS) are differentially expressed, we
can calculate

- the sensitivity or true positive rate (TPR), the proportion of
  actual positives that are correctly identified, in the protein list
  that we return

  $$TPR=\frac{TP}{\text{#actual positives}},$$

  here TP are the true positives in the list. The TPR is thus the
  fraction of ups proteins that we can recall.

- false discovery proportion (FPD): fraction of false positives in the
  protein list that we return:

  $$FPD=\frac{FP}{FP+TP},$$

  with FP the false positives. In our case the yeast proteins that are
  in our list.

Instead of only calculating that for the protein list that is returned
for the chosen FDR level, we can do this for all possible FDR cutoffs
so that we get an overview of the quality of the ranking of the
proteins in the protein list.

```{r sens}

fData(cptac_prot)$res %>%
                    dplyr::arrange(qval) %>%
                    dplyr::mutate(FDP = cumsum(!spike)/(1:length(spike)),
                                  TPR = cumsum(spike)/sum(spike)) %>% 
                    ggplot(aes(x = FDP, y = TPR)) +
                    geom_path() +
                    geom_point() + 
                    geom_vline(xintercept = 0.05, lty = 2)
```


# Further reading

- For additional material on `MSnbase` and MS/proteomics data
  manipualation, see
  [here](https://rawgit.com/lgatto/bioc-ms-prot/master/lab.html) and
  the [main `MSnbase`
  vignette](http://lgatto.github.io/MSnbase/articles/v01-MSnbase-demo.html).

- For additional `MSqRob` material, please see
  [here](https://statomics.github.io/pda/)

# Session information {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
